get-cabal-configuration.hs

```diff
@@ -302,39 +333,61 @@ isAllowedOption opt =
 
 dumpPackageDescription :: PackageDescription -> FilePath -> IO Sexp
 dumpPackageDescription pkgDesc projectDir = do
+#if defined(Cabal20) || defined(Cabal22) || defined(Cabal24)
+    (cabalNewDirs, cabalNewAutogen) <- getBuildDirectories (CabalNew $ package pkgDesc) pkgDesc projectDir
+#endif
     (cabalDirs, cabalAutogen) <- getBuildDirectories Cabal pkgDesc projectDir
     (stackDirs, stackAutogen) <- getBuildDirectories Stack pkgDesc projectDir
+#if defined(Cabal20) || defined(Cabal22) || defined(Cabal24)
+    let buildDirs   = cabalNewDirs ++ cabalDirs ++ stackDirs
+        autogenDirs = cabalNewAutogen ++ cabalAutogen ++ stackAutogen
+#else
     let buildDirs   = cabalDirs ++ stackDirs
         autogenDirs = cabalAutogen ++ stackAutogen
+#endif
     return $
         SList
             [ cons (sym "build-directories") (ordNub (map normalise buildDirs))
             , cons (sym "source-directories") sourceDirs
             , cons (sym "extensions") exts
             , cons (sym "languages") langs
+#if __GLASGOW_HASKELL__ < 800
+            -- No need to specify dependencies with package environments.
+            -- https://downloads.haskell.org/~ghc/8.0.2/docs/html/users_guide/packages.html#package-environments
```

Comment:
What if my project uses `stack`? The package environment file will be absent, or, even if present, would be irrelevant.

Moreover, I may use both cabal and stack for a single project and would like to be able to switch between the two. So, even though a package environment file will be present, it should have any effect only when relevant flycheck checker is used. Both Haskell and elisp in this project don't really know about that. The aim of the project is to get project configuration in a way that will make all further checks, e.g. plain ghc-based or stack based, work.

Reply:
I forgot about `stack` in this case. I need to test it myself with, even though I don't use it. I'm not sure whether it takes into account package environments or not.

In any case I don't think the package environments can be created with `stack`.

I removed the `dependencies` list because relying on package environment make it all much easier, but ideally, we should pass along the `dependencies` when there are not package environment set.

Reply:
To my knowledge stack neither creates nor user package environment files since everything is specified in the project's `stack.yaml` file. Beside, `stack.yaml` may specify a GHC version different from the one for which a package environment is available.

It is currently possible to switch between stack-based checker and cabal-based checker without doing reconfiguration. I would like to keep it that way.